### DISCUSSION TITLE
GraphQL working but UQL not working - help appricated!

### DISCUSSION THREAD

**Discussion by rogerjbos** [grafana/grafana-infinity-datasource#267] [Q&A] [answered]
Thanks for the great project.  Works great for GraphQL, but I am struggling with UQL.  My json looks like this:

{"query":{"dexActions":{"nodes":[{"id":"1518247-9","timestamp":"2022-02-15T15:29:18.326","volumeUSD":34.64837022927303},{"id":"1518256-29","timestamp":"2022-02-15T15:33:18.209","volumeUSD":35.57701107558557},{"id":"1518534-10","timestamp":"2022-02-15T17:04:00.415","volumeUSD":553.3645984610087}]}}}

With Type = GraphQL I am able to see the data using the following rows spec:

data.query.dexActions.nodes

So when I change the Type to UQL, I am trying the following:

parse-json
| scope "data.query.dexActions.nodes"
| project "id", "timestamp", "volumeUSD"

But it shows "400: Bad request"

Then I tried this:

parse-json
| project "data.query.dexActions.nodes.id", "data.query.dexActions.nodes.timestamp", "data.query.dexActions.nodes.volumeUSD"

same result. I tried a bunch of other variations, but always same "400: Bad request"  Any ideas how to correct my request?

In case it helps, my URL is 

https://api.subquery.network/sq/rogerjbos/acala-subql-modified

and the POST body for the GraphQL query is as follows:

```
{
  query {
    dexActions (filter: {timestamp: {greaterThanOrEqualTo: "2022-02-10"},
      type: {equalTo: "swap"}} first: 3) {
      nodes {
        id
        timestamp
        volumeUSD
      }
    }
  }
}
```

**Comment by yesoreyeram** [comment, marked as answer]
Alight.. I didn't thought about the graphql usecase for UQL. Glad you tried. Basically for graphql, we are parsing the body differently.

In case if you want to use UQL + graphql, you can try the following for now.
* Change the query type to UQL
* Wrap the body with the JSON `{ "query": "{ your graphql query goes here }"}
* Add a content type header

example : https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22url%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fapi.subquery.network%2Fsq%2Frogerjbos%2Facala-subql-modified%22,%22url_options%22:%7B%22method%22:%22POST%22,%22data%22:%22%7B%5Cn%5Ct%5C%22query%5C%22:%5C%22%7B%20query%20%7B%20dexActions%20(filter:%20%7Btimestamp:%20%7BgreaterThanOrEqualTo:%20%5C%5C%5C%222022-02-10%5C%5C%5C%22%7D,%20type:%20%7BequalTo:%20%5C%5C%5C%22swap%5C%5C%5C%22%7D%7D%20first:%203)%20%7B%20nodes%20%7B%20id%20timestamp%20volumeUSD%20%7D%20%7D%20%7D%20%7D%5C%22%5Cn%7D%22,%22headers%22:%5B%7B%22key%22:%22content-type%22,%22value%22:%22application%2Fjson%22%7D%5D%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22uql%22:%22parse-json%5Cn%7C%20scope%20%5C%22data.query.dexActions.nodes%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D

<img width="1787" alt="image" src="https://user-images.githubusercontent.com/153843/155196848-1464df2b-d47d-4c6a-be8e-2d601572b8ea.png">
<img width="618" alt="image" src="https://user-images.githubusercontent.com/153843/155197018-7fa190e0-77d3-447f-9624-a64c2dbb7309.png">
<img width="635" alt="image" src="https://user-images.githubusercontent.com/153843/155197035-7f955dc4-ee39-4361-a959-32dda94eabaf.png">

**Reply by rogerjbos** [reply]
Thank you so much for the answer and example. Really helpful. It is working perfect now.

**Reply by yesoreyeram** [reply]
Any specific reason you chose UQL over GraphQL? Any feedback will be helpful.

**Comment by rogerjbos** [comment]
Siriam,

Sorry for the late reply, just saw your email.  I chose the combination of
UQL + GraphQL because I needed to modify that data received from the
GraphQL call before graphing it in Grafana.  I am very new to GraphQL and
Grafana, so maybe I don't know all the options, but UQL was very helpful
for me to format the data as I needed it.

Thanks,

Roger
